### PR TITLE
[ADD][14.0] new module: base_order_by_related

### PR DESCRIPTION
**Pull request by thomaspaulb**
This module allows a search to order by a non-stored related field.

### CODE REVIEW

File: base_order_by_related/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2021 Akretion (https://www.akretion.com).
+# @author Kévin Roche <kevin.roche@akretion.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```
Comment (gjotten) [inline base_order_by_related/__manifest__.py]: @thomaspaulb needs different attribution methinks.

File: base_order_by_related/tests/fake_models.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# Copyright (C) 2022 Akretion (<http://www.akretion.com>).
+# @author Kévin Roche <kevin.roche@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
```
Comment (gjotten) [inline base_order_by_related/tests/fake_models.py]: @thomaspaulb here as well.

File: base_order_by_related/tests/test_order_by_related_mixin.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Copyright (C) 2022 Akretion (<http://www.akretion.com>).
+# @author Kévin Roche <kevin.roche@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
```
Comment (gjotten) [inline base_order_by_related/tests/test_order_by_related_mixin.py]: @thomaspaulb here as well.

File: base_order_by_related/models/order_by_related_mixin.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    @api.model
+    def _generate_order_by_inner(
+        self, alias, order_spec, query, reverse_direction=False, seen=None
+    ):
+        order_parts = order_spec.split(",")
+        order_fields = [p.strip().split(" ")[0].strip() for p in order_parts]
+        order_by_related = self._order_by_related or {}
+        if not any(f in order_by_related for f in order_fields):
+            return super(OrderByRelatedMixin, self)._generate_order_by_inner(
```
Comment (hbrunn) [inline base_order_by_related/models/order_by_related_mixin.py]:
```suggestion
            return super()._generate_order_by_inner(
```

File: base_order_by_related/models/order_by_related_mixin.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+class OrderByRelatedMixin(models.AbstractModel):
+    _name = "order.by.related.mixin"
+    _description = "Mixin to allow ordering by non-stored related fields"
+
+    @api.model
+    def _related_join_calc(self, alias, fname, query, real_fname):
+        model, field = self, self._fields[fname]
+        while field.related:
+            # retrieve the parent model where field is inherited from
+            related_model = self.env[field.related_field.model_name]
```
Comment (hbrunn) [inline base_order_by_related/models/order_by_related_mixin.py]:
```suggestion
            if field.related_field.model_name not in self.env:
                continue
            related_model = self.env[field.related_field.model_name]
```

for this kind of low-level stuff I'd suggest to be cautious about models/fields not existing due to a half loaded registry, also above with the field

File: base_order_by_related/readme/USAGE.rst [hunk 6]
```diff
@@ -0,0 +1,46 @@
+In our case we use this for the case of ordering common Odoo models by a non-stored computed field.
+
+This is normally not possible, but in this case it can be hacked in by creating a view such as::
```
Comment (hbrunn) [inline base_order_by_related/readme/USAGE.rst]: I found your example actually a bit too detailed. Don't we have some non-stored related field in the base module that you can use to make your point, same for tests?

File: base_order_by_related/models/order_by_related_mixin.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        else:
+            return '"%s"."%s"' % (alias, related_fname2)
+
+    @api.model
+    def _generate_order_by_inner(
+        self, alias, order_spec, query, reverse_direction=False, seen=None
+    ):
+        order_parts = order_spec.split(",")
+        order_fields = [p.strip().split(" ")[0].strip() for p in order_parts]
+        order_by_related = self._order_by_related or {}
```
Comment (hbrunn) [inline base_order_by_related/models/order_by_related_mixin.py]:
```suggestion
        order_by_related = getattr(self, '_order_by_related',  {})
```

**Review by hbrunn — APPROVED** [review]
code looks good to me, but personally I think I'd implement this as overrides in the `base` model, and calculate `_order_by_related` in `base#init` - that should make it more convenient for devs to use, don't you think?

**Comment by thomaspaulb** [comment]
> code looks good to me, but personally I think I'd implement this as overrides in the `base` model, and calculate `_order_by_related` in `base#init` - that should make it more convenient for devs to use, don't you think?

It would; I tried in two ways:

- Calculating `_order_by_related` in `base#init` for all non-stored related fields: this enables search and modifies the ORM call for a lot of problematic fields as well, so I think this should be an opt-in thing only
- Opt-in via a kwarg in `Field#init`: looked easy to do, but it will require another hook on the Field object which will complicate the module

So, I stick with `_order_by_related`, which has now simplified to being just a list.

**Review by hbrunn — APPROVED** [review]

**Review by gjotten — APPROVED** [review]

**Pull request merged**